### PR TITLE
Added Into trait for Error

### PR DESCRIPTION
**Pull request by dodomorandi**
This will be helpful with functions taking a raw error `i32` like
`std::io::Error::from_raw_os_error`.

### CODE REVIEW

**Comment by dodomorandi** [comment]
@lfairy Can we assume that getting `errno` without making anything else will always return 0 for every platform? If so, I can also add a simple test.

**Comment by lfairy** [comment]
> Can we assume that getting errno without making anything else will always return 0 for every platform? If so, I can also add a simple test.

I don't know. But since the correctness of our library doesn't depend on this behavior, I don't think it's worth testing for it anyway.

**Comment by lfairy** [comment]
Looks good to me, thanks!

Is it worth adding a

```rust
impl From<Errno> for io::Error
```

as well? Or do you think that would be too magical?

**Comment by dodomorandi** [comment]
I think that it would be a good idea, but I need your opinion. I was thinking about `std::io::Error::raw_os_error`: it returns an `Option<i32>`, therefore it should not be possible to obtain an `Errno` in this case. Maybe a better solution could be the unstable `TryFrom`. What do you think?

**Comment by lfairy** [comment]
I mean the other direction:

```rust
impl From<Errno> for io::Error {
    fn from(e: Errno) -> io::Error {
        io::Error::from_raw_os_error(e.0)
    }
}
```

I don't think that needs `TryFrom` 😛 

That will let you convert straight from a `Errno` to an `io::Error`, which I believe fits your original use case. It also allows for automatic conversion using `?` and `try!`.

**Comment by dodomorandi** [comment]
Woops, sorry, my fault! 😄

**TL;DR.** For me it is ok to implement this feature. I will send a commit ASAP. I have some doubts about possible misuses, but, IHMO, it is nothing that we can handle.

---

I still have some doubts. When a successful operation for _io_ operation using _std_ libs, we only get a `std::io::Error` object in case of error. At the same time, it is possible to use the `from_raw_os_error` with the error code `0`. This results is a _successful error_, as [seen here](https://play.rust-lang.org/?gist=9e365b133dcfcab998f2de9b918300e5&version=stable). At the same time this would be counter-intuitive, because an `Error` would be not expected in case of success. I imagine a situation like this:

```rust
use std::process::exit;
use std::ffi::CString;
use errno::errno;

fn main() {
    let fd = unsafe {
        libc::open(
            CString::new("/tmp/test.txt").unwrap().as_ptr(),
            libc::O_RDONLY,
        )
    };

    if fd != -1 { // <- WRONG! should be '=='
        let err: std::io::Error = errno().into();
        eprintln!("ERROR: {}", err);
        exit(-1);
    }

    unsafe {
        libc::close(fd);
    }
}
```

We have a typical _C bug_, where the `fd` is wrongly checked. It would be really nice to use a pattern to check against `errno()`, once it is converted to `std::io::Error`, but this brings us to some problems:

1. Even on linux (from the man page)
> the value of errno is never set to zero by any system call or library function

therefore a _successful_ operation would generally not set _errno_ to 0.
2. It is also said that
> a function that succeeds is allowed to change errno

which, IMHO, is the most difficult situation to handle, because one should check the _errno_ value even on success, but...
3. There is no `std::io::ErrorKind::Success`, so there is no way to use pattern matching to check for a _success error code_ (we get an `std::io::ErrorKind::Other`)

So, to summarize, it is not possible to just use `errno()` to safely check for successful unsafe operations. Or at least, I don't see how to do so.

Let's go back to our `From` impl.
IHMO using _longer expressions_ instead of a simple `.into()` will not prevent the occurrence of any bug. The fact that a _successful_ `std::io::Error` can be returned is something beyond our control. For me, it is ok to implement the feature exactly as you showed in your comment.

**Comment by lfairy** [comment]
@dodomorandi thanks for the thorough write-up! I agree with your conclusion, and will merge as-is.
